### PR TITLE
Usability improvements for interactive `init` task

### PR DESCRIPTION
**Pull request by adammurdoch**
### Context

This PR makes a number of improvements to the `init` task when used interactively:
- Rework how the "type" is displayed, and default to generating an application. This is more likely to be useful than defaulting to "basic"
- Remove spurious warning when using `init` with configuration caching enabled.
- Reorder the implementation languages by descending likelihood of being selected. The default is still "Java"'
- Default Java version to 21, the most recent LTS, rather than whatever happens to be in the user's PATH.
- Prompt the user again if they enter an invalid Java version, instead of failing with an exception.
- Move the "Java version" question up to just after the implementation language, to group related questions. Previously it was the last question asked.
- Move the "project name" question earlier.
- Rework how the modularization options are displayed, to (hopefully) clarify what they do.
- Allow the user to enter 'y' or 'n' in response to a yes/no question, instead of requiring them to type 'yes' or 'no'. The build scan "accept license?" question still requires an explicit "yes" or "no".
- Some internal refactors, eg to remove the "none" language and explicitly model the Swift and C++ test frameworks.

#### Build type question

New output:

```
Select type of build to generate:
  1: Application
  2: Library
  3: Gradle plugin
  4: Basic (build structure only)
Enter selection (default: Application) [1..4] 
```

Gradle 8.5:

```
Select type of project to generate:
  1: basic
  2: application
  3: library
  4: Gradle plugin
Enter selection (default: basic) [1..4]
```

#### Implementation language question

New output:

```
Select implementation language:
  1: Java
  2: Kotlin
  3: Groovy
  4: Scala
  5: C++
  6: Swift
Enter selection (default: Java) [1..6] 
```

Gradle 8.5

```
Select implementation language:
  1: C++
  2: Groovy
  3: Java
  4: Kotlin
  5: Scala
  6: Swift
Enter selection (default: Java) [1..6] 
```

#### Modularization question

New output:

```
Select application structure:
  1: Single application project
  2: Application and library project
Enter selection (default: Single application project) [1..2] 
```

Gradle 8.5:

```
Generate multiple subprojects for application? (default: no) [yes, no] 
```

#### Java version question

New output:

```
Enter target Java version (min: 7, default: 21): 
```

Gradle 8.5

```
Enter target version of Java (min. 7) (default: 11):
```

#### Full interaction

Here's an example of the new output with default answers to all questions:

```
Select type of build to generate:
  1: Application
  2: Library
  3: Gradle plugin
  4: Basic (build structure only)
Enter selection (default: Application) [1..4] 

Select implementation language:
  1: Java
  2: Kotlin
  3: Groovy
  4: Scala
  5: C++
  6: Swift
Enter selection (default: Java) [1..6] 

Enter target Java version (min: 7, default: 21): 

Project name (default: test): 

Select application structure:
  1: Single application project
  2: Application and library project
Enter selection (default: Single application project) [1..2] 

Select build script DSL:
  1: Kotlin
  2: Groovy
Enter selection (default: Kotlin) [1..2] 

Select test framework:
  1: JUnit 4
  2: TestNG
  3: Spock
  4: JUnit Jupiter
Enter selection (default: JUnit Jupiter) [1..4] 

Generate build using new APIs and behavior (some features may change in the next minor release)? (default: no) [yes, no] 

> Task :init
To learn more about Gradle by exploring our Samples at https://docs.gradle.org/8.7-20240123180757+0000/samples/sample_building_java_applications.html

BUILD SUCCESSFUL in 23s
1 actionable task: 1 executed
```

Here's the same build with Gradle 8.5:

```
Select type of project to generate:
  1: basic
  2: application
  3: library
  4: Gradle plugin
Enter selection (default: basic) [1..4] 2

Select implementation language:
  1: C++
  2: Groovy
  3: Java
  4: Kotlin
  5: Scala
  6: Swift
Enter selection (default: Java) [1..6] 

Generate multiple subprojects for application? (default: no) [yes, no] 
Select build script DSL:
  1: Kotlin
  2: Groovy
Enter selection (default: Kotlin) [1..2] 

Select test framework:
  1: JUnit 4
  2: TestNG
  3: Spock
  4: JUnit Jupiter
Enter selection (default: JUnit Jupiter) [1..4] 

Project name (default: test): 
Source package (default: test): 
Enter target version of Java (min. 7) (default: 11): 21
Generate build using new APIs and behavior (some features may change in the next minor release)? (default: no) [yes, no] 

> Task :init
To learn more about Gradle by exploring our Samples at https://docs.gradle.org/8.5/samples/sample_building_java_applications.html

BUILD SUCCESSFUL in 20s
2 actionable tasks: 2 executed
```

### Contributor Checklist
- [ ] [Review Contribution Guidelines](https://github.com/gradle/gradle/blob/master/CONTRIBUTING.md)
- [ ] Make sure that all commits are [signed off](https://git-scm.com/docs/git-commit#Documentation/git-commit.txt---signoff) to indicate that you agree to the terms of [Developer Certificate of Origin](https://developercertificate.org/).
- [ ] Make sure all contributed code can be distributed under the terms of the [Apache License 2.0](https://github.com/gradle/gradle/blob/master/LICENSE), e.g. the code was written by yourself or the original code is licensed under [a license compatible to Apache License 2.0](https://apache.org/legal/resolved.html).
- [ ] Check ["Allow edit from maintainers" option](https://help.github.com/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork/) in pull request so that additional changes can be pushed by Gradle team
- [ ] Provide integration tests (under `<subproject>/src/integTest`) to verify changes from a user perspective
- [ ] Provide unit tests (under `<subproject>/src/test`) to verify logic
- [ ] Update User Guide, DSL Reference, and Javadoc for public-facing changes
- [ ] Ensure that tests pass sanity check: `./gradlew sanityCheck`
- [ ] Ensure that tests pass locally: `./gradlew <changed-subproject>:quickTest`

### Reviewing cheatsheet

Before merging the PR, comments starting with 
- ❌ ❓**must** be fixed
- 🤔 💅 **should** be fixed
- 💭 **may** be fixed
- 🎉 celebrate happy things

### CODE REVIEW

File: platforms/software/build-init/src/integTest/groovy/org/gradle/buildinit/plugins/BuildInitInteractiveIntegrationTest.groovy [hunk 1]
```diff
@@ -18,20 +18,30 @@ package org.gradle.buildinit.plugins
...
-    String projectTypePrompt = "Select type of project to generate:"
-    String dslPrompt = "Select build script DSL:"
-    String incubatingPrompt = "Generate build using new APIs and behavior (some features may change in the next minor release)?"
-    String basicType = "1: basic"
-    String projectNamePrompt = "Project name (default: some-thing)"
-    String convertMavenBuildPrompt = "Found a Maven build. Generate a Gradle build from this?"
-    List<String> languageSelectionOptions = ["Select implementation language:", "1: C++", "2: Groovy", "3: Java", "4: Kotlin", "5: Scala", "6: Swift"]
+    def buildTypePrompt = "Select type of build to generate:"
+    def dslPrompt = "Select build script DSL:"
+    def incubatingPrompt = "Generate build using new APIs and behavior (some features may change in the next minor release)?"
```
Comment (alllex) [inline platforms/software/build-init/src/integTest/groovy/org/gradle/buildinit/plugins/BuildInitInteractiveIntegrationTest.groovy:29]: I think this question is very cumbersome for the users as well. Without the context, it begs the question of why would new APIs change in the minor release?

I think it would be much better to just say "Use incubating Gradle APIs?". For those who know, the question is short and clear. For those who don't know, the word "incubating" gives a hint, and the first Google result for "Gradle incubating" already gives the full answer. It also is better connected to the CLI option `--incubating` that corresponds to this question.

@lkasso, could you give your input here as well?

File: platforms/software/build-init/src/main/java/org/gradle/buildinit/plugins/BuildInitPlugin.java [hunk 2]
```diff
@@ -72,7 +72,12 @@ public void apply(Project project) {
     }
 
     private static Provider<Boolean> getCommentsProperty(Project project) {
-        return project.getProviders().gradleProperty(COMMENTS_PROPERTY).map(Boolean::parseBoolean);
+        return project.getProviders().gradleProperty(COMMENTS_PROPERTY).map(new Transformer<Boolean, String>() {
```
Comment (alllex) [inline platforms/software/build-init/src/main/java/org/gradle/buildinit/plugins/BuildInitPlugin.java:75]: ❔ Why is this change required?

File: platforms/software/build-init/src/main/java/org/gradle/buildinit/plugins/internal/AbstractBuildGenerator.java [hunk 3]
```diff
@@ -0,0 +1,54 @@
+/*
+ * Copyright 2018 the original author or authors.
```
Comment (alllex) [inline platforms/software/build-init/src/main/java/org/gradle/buildinit/plugins/internal/AbstractBuildGenerator.java:2]: ❓ Should the new files use the current year for copyright?

**Review by alllex — COMMENTED** [review]
This is a lot of great changes!

I left some comments, though I made it only half way through the PR. It would have been easier if the PR was split into a couple 😉

File: platforms/software/build-init/src/integTest/groovy/org/gradle/buildinit/plugins/BuildInitInteractiveIntegrationTest.groovy [hunk 4]
```diff
@@ -18,20 +18,30 @@ package org.gradle.buildinit.plugins
...
 class BuildInitInteractiveIntegrationTest extends AbstractInitIntegrationSpec {
 
-    String projectTypePrompt = "Select type of project to generate:"
-    String dslPrompt = "Select build script DSL:"
-    String incubatingPrompt = "Generate build using new APIs and behavior (some features may change in the next minor release)?"
-    String basicType = "1: basic"
-    String projectNamePrompt = "Project name (default: some-thing)"
-    String convertMavenBuildPrompt = "Found a Maven build. Generate a Gradle build from this?"
-    List<String> languageSelectionOptions = ["Select implementation language:", "1: C++", "2: Groovy", "3: Java", "4: Kotlin", "5: Scala", "6: Swift"]
+    def buildTypePrompt = "Select type of build to generate:"
```
[the remark this answers is not in the document]
Comment (lkasso) [reply platforms/software/build-init/src/integTest/groovy/org/gradle/buildinit/plugins/BuildInitInteractiveIntegrationTest.groovy:27]: @alllex, this process is a mechanism of sorts so I actually think `generate` is a good choice.

File: platforms/software/build-init/src/integTest/groovy/org/gradle/buildinit/plugins/BuildInitInteractiveIntegrationTest.groovy (hunk 1 above)
Comment (lkasso) [reply platforms/software/build-init/src/integTest/groovy/org/gradle/buildinit/plugins/BuildInitInteractiveIntegrationTest.groovy:29]: I never understood what it meant and had to ask @big-guy when I first used it.
I agree with @alllex, `Use incubating Gradle APIs?` or similar is a much better choice.

File: platforms/software/build-init/src/integTest/groovy/org/gradle/buildinit/plugins/BuildInitInteractiveIntegrationTest.groovy [hunk 5]
```diff
@@ -119,48 +129,50 @@ class BuildInitInteractiveIntegrationTest extends AbstractInitIntegrationSpec {
...
+        handle.stdinPipe.write(TextUtil.platformLineSeparator.bytes)
 
-        // Select 'junit'
+        // Select 'Single project'
         ConcurrentTestUtil.poll(60) {
-            assert handle.standardOutput.contains("Select test framework:")
-            assert handle.standardOutput.contains("1: JUnit 4")
+            assert handle.standardOutput.contains("Select application structure:")
+            assert handle.standardOutput.contains("1: Single application project")
+            assert handle.standardOutput.contains("2: Application and library project")
```
Comment (lkasso) [inline platforms/software/build-init/src/integTest/groovy/org/gradle/buildinit/plugins/BuildInitInteractiveIntegrationTest.groovy:160]: The options for `Select application structure:` don't make sense to me. As a new user, I wouldn't know the difference between options 1 and 2.

File: platforms/software/build-init/src/main/java/org/gradle/buildinit/plugins/internal/modifiers/BuildInitTestFramework.java [hunk 6]
```diff
@@ -27,7 +27,9 @@ public enum BuildInitTestFramework implements WithIdentifier {
     SPOCK("Spock"),
     KOTLINTEST("kotlin.test"),
     SCALATEST("ScalaTest"),
-    JUNIT_JUPITER("JUnit Jupiter");
+    JUNIT_JUPITER("JUnit Jupiter"),
+    XCTest("XCTest"),
```
Comment (alllex) [inline platforms/software/build-init/src/main/java/org/gradle/buildinit/plugins/internal/modifiers/BuildInitTestFramework.java:31]: ❓ Is it intentional that the enum value naming here is not SNAKE_CASE?

File: subprojects/core/src/main/java/org/gradle/api/internal/tasks/userinput/DefaultUserInputHandler.java [hunk 7]
```diff
@@ -54,23 +57,42 @@ public Boolean askYesNoQuestion(String question) {
...
         builder.append(" (default: ");
-        builder.append(defaultValue ? "yes" : "no");
+        String defaultString = defaultValue ? "yes" : "no";
+        builder.append(defaultString);
         builder.append(") [");
         builder.append(StringUtils.join(YES_NO_CHOICES, ", "));
         builder.append("] ");
-        return prompt(builder.toString(), defaultValue, new BooleanParser());
+        return prompt(builder.toString(), defaultValue, value -> {
+            if (LENIENT_YES_NO_CHOICES.contains(value.toLowerCase(Locale.US))) {
```
Comment (alllex) [inline subprojects/core/src/main/java/org/gradle/api/internal/tasks/userinput/DefaultUserInputHandler.java:80]: 👍  Thanks for making it case lenient as well

File: subprojects/core/src/test/groovy/org/gradle/api/internal/tasks/userinput/DefaultUserInputHandlerTest.groovy [hunk 8]
```diff
@@ -89,11 +88,22 @@ class DefaultUserInputHandlerTest extends Specification {
 
         and:
         input == false
+
+        where:
+        invalidInput | _
```
Comment (alllex) [inline subprojects/core/src/test/groovy/org/gradle/api/internal/tasks/userinput/DefaultUserInputHandlerTest.groovy:93]: 💅  Consider using `invalidInput << ['', 'bla', 'y', ...]`

File: subprojects/core/src/test/groovy/org/gradle/api/internal/tasks/userinput/DefaultUserInputHandlerTest.groovy [hunk 9]
```diff
@@ -132,23 +150,29 @@ class DefaultUserInputHandlerTest extends Specification {
...
         1 * userInputReader.readInput() >> 'no'
         1 * outputEventBroadcaster.onOutput(_) >> { PromptOutputEvent event -> assert event.prompt == TextUtil.platformLineSeparator }
         1 * outputEventBroadcaster.onOutput(_ as UserInputResumeEvent)
         0 * outputEventBroadcaster._
 
         and:
         input == false
+
+        where:
+        invalidInput | _
```
Comment (alllex) [inline subprojects/core/src/test/groovy/org/gradle/api/internal/tasks/userinput/DefaultUserInputHandlerTest.groovy:172]: 💅  The second column is unused, but contains values. Could use `invalidInput << [...]` instead

File: subprojects/core/src/main/java/org/gradle/api/internal/tasks/userinput/UserInputHandler.java [hunk 10]
```diff
@@ -20,44 +20,94 @@
...
-     * Prompts the user with a yes/no question and returns the answer.
+     * Prompts the user with a boolean question and returns the answer.
      *
      * @param question The text of the question.
-     * @return the answer or the given default if not connected to a user console.
+     * @param defaultValue The option to present to the user as the default choice, and the value to use when not connected to a console
+     * @return the answer or the given default if not connected to a console.
      */
-    boolean askYesNoQuestion(String question, boolean defaultValue);
+    boolean askBooleanQuestion(String question, boolean defaultValue);
```
Comment (alllex) [inline subprojects/core/src/main/java/org/gradle/api/internal/tasks/userinput/UserInputHandler.java:43]: 🤔 Since the semantics changed, it makes sense to rename this method, but I think the naming can be improved.

From the user interaction POV, it's still a yes/no question, but "boolean" name does not imply this. Also, having an `askIntQuestion` creates an impression that this question is actually true/false question, which it is not.

What about `askLenientYesNoQuestion()` or something similar?

File: platforms/software/build-init/src/main/java/org/gradle/buildinit/plugins/internal/ProjectLayoutSetupRegistry.java [hunk 11]
```diff
@@ -54,16 +54,43 @@ public TemplateOperationFactory getTemplateOperationFactory() {
...
     public List<ComponentType> getComponentTypes() {
-        return ImmutableList.copyOf(ComponentType.values());
+        return Arrays.asList(ComponentType.values());
+    }
+
+    /**
+     * Returns the default component type to use for interactive initialization.
+     */
+    public ComponentType getDefaultComponentType() {
+        return getComponentTypes().get(0);
```
Comment (alllex) [inline platforms/software/build-init/src/main/java/org/gradle/buildinit/plugins/internal/ProjectLayoutSetupRegistry.java:85]: ❔ Would it better to be explicit here with `ComponentType.APPLICATION`, since it's a enum?

**Review by alllex — COMMENTED** [review]
I went through the whole thing this time. Left more comments

File: platforms/software/build-init/src/integTest/groovy/org/gradle/buildinit/plugins/BuildInitInteractiveIntegrationTest.groovy (hunk 1 above)
Comment (adammurdoch) [reply platforms/software/build-init/src/integTest/groovy/org/gradle/buildinit/plugins/BuildInitInteractiveIntegrationTest.groovy:29]: For sure. I didn't change that prompt in this PR (that's a test refactoring). It's on the list for later.

File: platforms/software/build-init/src/main/java/org/gradle/buildinit/plugins/BuildInitPlugin.java (hunk 2 above)
Comment (adammurdoch) [reply platforms/software/build-init/src/main/java/org/gradle/buildinit/plugins/BuildInitPlugin.java:75]: It's to make the init task CC compatible. We still can't use lambdas in the core plugins.

File: platforms/software/build-init/src/main/java/org/gradle/buildinit/plugins/internal/AbstractBuildGenerator.java (hunk 3 above)
Comment (adammurdoch) [reply platforms/software/build-init/src/main/java/org/gradle/buildinit/plugins/internal/AbstractBuildGenerator.java:2]: It should indeed. Not sure what happened there.

File: platforms/software/build-init/src/integTest/groovy/org/gradle/buildinit/plugins/BuildInitInteractiveIntegrationTest.groovy (hunk 5 above)
Comment (adammurdoch) [reply platforms/software/build-init/src/integTest/groovy/org/gradle/buildinit/plugins/BuildInitInteractiveIntegrationTest.groovy:160]: For this PR, I think the prompt just needs to be better than it was before ("Generate multiple subprojects for application? (default: no) [yes, no] "), which I think it is.

I'll see if I can tweak it to make it clearer. We also have the option to say "if you don't know, use the default".

File: platforms/software/build-init/src/main/java/org/gradle/buildinit/plugins/internal/modifiers/BuildInitTestFramework.java (hunk 6 above)
Comment (adammurdoch) [reply platforms/software/build-init/src/main/java/org/gradle/buildinit/plugins/internal/modifiers/BuildInitTestFramework.java:31]: No, it's not.

File: subprojects/core/src/test/groovy/org/gradle/api/internal/tasks/userinput/DefaultUserInputHandlerTest.groovy (hunk 8 above)
Comment (adammurdoch) [reply subprojects/core/src/test/groovy/org/gradle/api/internal/tasks/userinput/DefaultUserInputHandlerTest.groovy:93]: I did. I tend to find this layout clearer so that I can quickly scan the cases more easily.

**Comment by adammurdoch** [comment]
Thanks for the review @alllex. Sorry it's all in one giant PR, but I'm on holidays so I wanted to minimize the number of times I had to go to slack.

File: subprojects/core/src/main/java/org/gradle/api/internal/tasks/userinput/UserInputHandler.java (hunk 10 above)
Comment (adammurdoch) [reply subprojects/core/src/main/java/org/gradle/api/internal/tasks/userinput/UserInputHandler.java:43]: I'm not sure. I don't really want the name to imply that the user has to type 'yes' or 'no'. The client may be a UI with a checkbox, for example.

The intent of other method is quite different, as it requires the user to explicitly type 'yes' to agree to something (eg license terms, etc), so I don't want to use a similar name that implies similar semantics.

**Comment by adammurdoch** [comment]
@bot-gradle test and merge

**Comment by adammurdoch** [comment]
@bot-gradle test and merge

File: platforms/software/build-init/src/main/java/org/gradle/buildinit/plugins/BuildInitPlugin.java (hunk 2 above)
Comment (alllex) [reply platforms/software/build-init/src/main/java/org/gradle/buildinit/plugins/BuildInitPlugin.java:75]: Then you could use a helper method instead of an anonymous class: `SerializableLambdas.transformer(Boolean::parseBoolean)`

Comment (alllex) [reply platforms/software/build-init/src/main/java/org/gradle/buildinit/plugins/BuildInitPlugin.java:75]: I now realize the helper is an internal API, and we probably want to limit the usage of internal APIs even in the core plugin implementations.

However, the current requirement for wrapping everything into anonymous classes seems excessive just to make things CC compatible. Especially so, in the cases like this.

File: platforms/software/build-init/src/integTest/groovy/org/gradle/buildinit/plugins/AbstractJvmLibraryInitIntegrationSpec.groovy [hunk 12]
```diff
@@ -23,7 +23,10 @@ abstract class AbstractJvmLibraryInitIntegrationSpec extends AbstractInitIntegra
 
     def setup() {
         executer.beforeExecute { e ->
+            // The init task defaults to Java 21, which may not be the current JVM.
```
[the remark this answers is not in the document]
Comment (adammurdoch) [reply platforms/software/build-init/src/integTest/groovy/org/gradle/buildinit/plugins/AbstractJvmLibraryInitIntegrationSpec.groovy]: Updated this comment

File: platforms/software/build-init/src/main/java/org/gradle/buildinit/plugins/internal/AbstractBuildGenerator.java (hunk 3 above)
Comment (adammurdoch) [reply platforms/software/build-init/src/main/java/org/gradle/buildinit/plugins/internal/AbstractBuildGenerator.java]: Fixed

File: platforms/software/build-init/src/main/java/org/gradle/buildinit/plugins/internal/modifiers/BuildInitTestFramework.java (hunk 6 above)
Comment (adammurdoch) [reply platforms/software/build-init/src/main/java/org/gradle/buildinit/plugins/internal/modifiers/BuildInitTestFramework.java]: Fixed

File: platforms/software/build-init/src/main/java/org/gradle/buildinit/plugins/internal/ProjectLayoutSetupRegistry.java (hunk 11 above)
Comment (adammurdoch) [reply platforms/software/build-init/src/main/java/org/gradle/buildinit/plugins/internal/ProjectLayoutSetupRegistry.java]: I've changed this
